module/VuFind/src/VuFind/Resolver/Driver/Jop.php

```diff
@@ -187,58 +198,64 @@ public function getResolverUrl($openURL)
+        // the JOP link resolver returns unstyled XML which is not helpful for the
         // user
         return false;
     }
 
     /**
-     * Downgrade an OpenURL from v1.0 to v0.1 for compatibility with EZB.
+     * Downgrade an OpenURL from v1.0 to v0.1 for compatibility with JOP.
      *
      * @param array $parsed Array of parameters parsed from the OpenURL.
      *
-     * @return string       EZB-compatible v0.1 OpenURL
+     * @return string       JOP-compatible v0.1 OpenURL
      */
     protected function downgradeOpenUrl($parsed)
     {
         $downgraded = [];
 
-        // we need 'genre' but only the values
-        // article or journal are allowed...
-        $downgraded[] = "genre=article";
+        // prepare content for downgrading
+        // resolver only accepts date formats YYYY, YYYY-MM, and YYYY-MM-DD
+        // in case we have a date in another format, drop the date information
+        if (isset($parsed['rft.date'])
+            && !preg_match('/^\d{4}(-\d\d(-\d\d)?)?$/', $parsed['rft.date'])
+        ) {
+            unset($parsed['rft.date']);
+        }
+
+        $map = [
+            'rfr_id' => 'sid',
+            'rft.date' => 'date',
+            'rft.issn' => 'issn',
+            'rft.isbn' => 'isbn', // isbn is supported as of 12/2021
+            'rft.volume' => 'volume',
+            'rft.issue' => 'issue',
+            'rft.spage' => 'spage',
+            'rft.pages' => 'pages',
+        ];
 
         // ignore all other parameters
         foreach ($parsed as $key => $value) {
             // exclude empty parameters
             if (isset($value) && $value !== '') {
-                if ($key == 'rfr_id') {
-                    $newKey = 'sid';
-                } elseif ($key == 'rft.date') {
-                    $newKey = 'date';
-                } elseif ($key == 'rft.issn') {
-                    $newKey = 'issn';
-                } elseif ($key == 'rft.volume') {
-                    $newKey = 'volume';
-                } elseif ($key == 'rft.issue') {
-                    $newKey = 'issue';
-                } elseif ($key == 'rft.spage') {
-                    $newKey = 'spage';
-                } elseif ($key == 'rft.pages') {
-                    $newKey = 'pages';
-                } else {
-                    $newKey = false;
-                }
-                if ($newKey !== false) {
-                    $downgraded[] = "$newKey=$value";
+                if (isset($map[$key])) {
+                    $downgraded[] = "{$map[$key]}=$value";
+                } elseif (in_array($key, $map)) {
+                    $downgraded[] = "$key=$value";
                 }
             }
         }
+        if (!empty($downgraded)) {
+            // we need 'genre' but only the values
+            // article or journal are allowed...
+            return "genre=article&" . implode('&', $downgraded);
+        }
 
-        return implode('&', $downgraded);
+        return implode('&', $parsed);
```

Comment:
Is there a reason we pass back the original settings if we were unable to find anything to downgrade? This seems like a behavior change from before, and I'm not sure that it will end up with good results. If there's a reason for it, maybe a comment would help explain why.